vpn/client.go

```diff
@@ -0,0 +1,180 @@
+package vpn
+
+import (
+	"context"
+	"net/http"
+	"net/netip"
+	"net/url"
+
+	"golang.org/x/xerrors"
+	"nhooyr.io/websocket"
+	"tailscale.com/net/dns"
+	"tailscale.com/wgengine/router"
+
+	"github.com/tailscale/wireguard-go/tun"
+
+	"cdr.dev/slog"
+	"github.com/coder/coder/v2/codersdk"
+	"github.com/coder/coder/v2/codersdk/workspacesdk"
+	"github.com/coder/coder/v2/tailnet"
+	"github.com/coder/coder/v2/tailnet/proto"
+	"github.com/coder/quartz"
+)
+
+type Conn interface {
+	CurrentWorkspaceState() *proto.WorkspaceUpdate
```

Comment:
the Tailnet protocol doesn't include things like domain names, last handshake, or IP addresses that we're supposed to report out in the VPN protocol. So I think we'll need to define new golang types that include that info, so that the workspace updates controller can provide it to the VPN.